### PR TITLE
feat(flux2klein): Faster Loading + Compliation

### PR DESCRIPTION
**Pull request by amepas**
**Summary**
Significantly improving the model loading and compilation times for the Flux Klein models. This enables faster start-up and lets us iterate faster on model improvements. We also include bug fixes in Flux2.Klein implementation (handling of batch-sizes, making sure all code goes through NNX modules rather than linen).

**Optimizations**
1. Loading weights directly into bfloat16: Previous benchmarking of implementation required us to be able to run 
the model in fp32, but this is not needed anymore. 
2. Parallelizing compilation of Qwen3 text encoder, denoising steps, vae
3. Use only 1 step of the denoiser instead of 4
4. Fixes to VAE:  replaced jax.image.resize(hidden_states, shape=(batch, height * 2, width * 2, channels), method="nearest") with 6D broadcast + reshape 2x upsampling . This saved a lot of time in compilation 
6. Latent tensor sharding: Ensuring the sharding is maintained across the layers so that there is no re-compilation or re-sharding that needs to be triggered when we do a warm-up pass. 

None of these changes affect the model implementation, so the outputs are still identical.

**Results**
On a v6-4, we see the following speedups (BS=8, fsdp) for the 9B model:
- Warm Model Loading: 154s -> 12s 
- XLA compilation/warmup-pass: 61s -> 11s
- Inference: unchanged

On a v7-8:
-Warm model loading: 540s -> 140s (for cold loads, use a Hyperdisk mount) 
-XLA compilation/warmup-pass: 82s -> 49s
-Inference unchanged

So, on the v6-4: We bring the loading+compilation time  **down to <30s**. On a v7-8, **<4 min**. The main bottleneck is the hardware constraint of how fast it can load the model. 

**Correctness**
Visual results are high-quality, SSIM changes from 1.00 to 0.9 (still passing all parity checks against PyTorch). Slight changes to SSIM are expected given that using NNX will lead to different graphs than linen, and thus cause different handling of floats. Reference images updated to maintain 1.00 sim on the repo.

### CODE REVIEW

File: src/maxdiffusion/generate_flux2klein.py [hunk 1]
```diff
@@ -184,8 +201,13 @@ def main(argv):
...
 
-    max_logging.log(f"Resolving snapshot directory for model '{repo_id}' from HF Hub...")
-    snapshot_dir = snapshot_download(repo_id=repo_id)
+    rev = getattr(config, "revision", None)
+    if not rev or rev == "refs/pr/95":
+      rev = "main"
+    try:
+      snapshot_dir = snapshot_download(repo_id=repo_id, revision=rev, local_files_only=True)
+    except Exception:
+      snapshot_dir = snapshot_download(repo_id=repo_id, local_files_only=True)
```
[the remark this answers is not in the document]
Comment (amepas) [reply src/maxdiffusion/generate_flux2klein.py]: Made a new commit that resolves both of these comments

File: src/maxdiffusion/tests/generate_flux2klein_smoke_test.py [hunk 2]
```diff
@@ -117,7 +117,7 @@ def test_flux2klein_9b_smoke(self):
     self.assertEqual(base_image.shape, test_image.shape)
     ssim_compare = ssim(base_image, test_image, channel_axis=-1, data_range=255)
     print(f"\n[SMOKE TEST 9B] SSIM Score: {ssim_compare:.6f}")
-    self.assertGreaterEqual(ssim_compare, 0.80)
+    self.assertGreaterEqual(ssim_compare, 0.78)
```
Comment (chandrasekhard2) [inline src/maxdiffusion/tests/generate_flux2klein_smoke_test.py]: Why did we reduce the threshold here?

File: src/maxdiffusion/models/flux/util.py [hunk 3]
```diff
@@ -398,19 +398,22 @@ def cast_dict_to_bfloat16_inplace(d, device=None, exclude_keywords=None, parent_
...
-      if hasattr(d[k], "block_until_ready"):
-        d[k].block_until_ready()
-      del v
-      gc.collect()
+      if v.dtype != target_dtype:
+        d[k] = v.astype(target_dtype)
+        if hasattr(d[k], "block_until_ready"):
+          d[k].block_until_ready()
+        del v
+        gc.collect()
```
Comment (chandrasekhard2) [inline src/maxdiffusion/models/flux/util.py:407]: is it good to move this out of for loop?

**Review by chandrasekhard2 — CHANGES_REQUESTED** [review]

File: src/maxdiffusion/tests/generate_flux2klein_smoke_test.py (hunk 2 above)
Comment (amepas) [reply src/maxdiffusion/tests/generate_flux2klein_smoke_test.py]: I think the smoke test baseline image is a bit outdated, so the SSIM was about 0.805. It went to 0.795 so I just adjusted the threshold because visually the images are identical. I can update the baseline image as well

File: src/maxdiffusion/models/flux/util.py (hunk 3 above)
Comment (amepas) [reply src/maxdiffusion/models/flux/util.py:407]: Its a bit slower to have it within the for loop, but I think its safer to avoid OOM. By keeping it in the for loop the memory does not spike as much during loading (it would hit 54GB of peak memory vs only 18GB otw).

File: src/maxdiffusion/tests/generate_flux2klein_smoke_test.py (hunk 2 above)
Comment (amepas) [reply src/maxdiffusion/tests/generate_flux2klein_smoke_test.py]: Stale reference images replaced, SSIM threshold restored back to 0.8

File: src/maxdiffusion/models/flux/util.py (hunk 3 above)
Comment (amepas) [reply src/maxdiffusion/models/flux/util.py:407]: With this code within the for loop, model loading takes <20s on v6-4. Moving the code out of the for loop does not noticeably affect loading time but keeps the peak memory much lower. Marking this as resolved, but happy to discuss further if needed

File: src/maxdiffusion/pipelines/flux/flux2klein_pipeline.py [hunk 4]
```diff
@@ -199,14 +319,28 @@ def __call__(
...
+
       # ---------------------------------------------------------------------
       # PHASE A: Encode Prompt (Qwen3)
       # ---------------------------------------------------------------------
+      if prompts is None:
+        prompts = ["A dog running in a field with butterflies and tall grass"]
+      elif isinstance(prompts, str):
+        prompts = [prompts]
+
       print(f"{host_prefix} [PHASE A] Encoding {len(prompts)} prompt(s) using JAX Qwen3 on TPU...", flush=True)
```
Comment (eltsai) [inline src/maxdiffusion/pipelines/flux/flux2klein_pipeline.py]: please change all print statements to max_logging

File: src/maxdiffusion/generate_flux2klein.py [hunk 5]
```diff
@@ -528,24 +568,101 @@ def unbox_fn(x):
...
     )
-    main_time = (
-        main_trace.get("prompt_encoding", 0.0) + main_trace.get("denoise_loop", 0.0) + main_trace.get("vae_decode", 0.0)
+    warmup_time = (
+        warmup_trace.get("prompt_encoding", 0.0)
+        + warmup_trace.get("denoise_loop", 0.0)
+        + warmup_trace.get("vae_decode", 0.0)
     )
 
+    num_reps = int(getattr(config, "num_reps", 1))
```
Comment (eltsai) [inline src/maxdiffusion/generate_flux2klein.py:548]: I think this variable is not defined in the yml file?

File: src/maxdiffusion/pipelines/flux/flux2klein_pipeline.py [hunk 6]
```diff
@@ -71,6 +70,29 @@ def __init__(
...
+            config, "pretrained_model_name_or_path", ""
+        )
+        hf_home = os.environ.get("HF_HOME", os.path.expanduser("~/.cache/huggingface"))
+        repo_cache = os.path.join(
+            hf_home, "hub", f"models--{getattr(config, 'pretrained_model_name_or_path', '').replace('/', '--')}", "snapshots"
+        )
+        if os.path.exists(repo_cache) and os.listdir(repo_cache):
+          tokenizer_path = os.path.join(repo_cache, os.listdir(repo_cache)[0])
+
+        from transformers import Qwen2TokenizerFast
```
Comment (eltsai) [inline src/maxdiffusion/pipelines/flux/flux2klein_pipeline.py]: The tokenizer is also initialized in `__init__`

Comment (eltsai) [reply src/maxdiffusion/pipelines/flux/flux2klein_pipeline.py]: I am not sure how big a diff for measured e2e time it would cause to init tokenizer here (probably minimal), but it's better to do this instead:

keep the `__init__` resolution but route it through `register_modules(tokenizer=tokenizer)` after the fallback (rather than assigning the attribute directly), then have `__call__` use self.tokenizer instead of constructing its own.

File: src/maxdiffusion/pipelines/flux/flux2klein_pipeline.py [hunk 7]
```diff
@@ -199,14 +319,28 @@ def __call__(
...
+          return x
+        if hasattr(sharding, "is_fully_addressable") and sharding.is_fully_addressable:
+          return jax.device_put(x, sharding)
+        return device_put_replicated(x, sharding)
+
       # ---------------------------------------------------------------------
       # PHASE A: Encode Prompt (Qwen3)
       # ---------------------------------------------------------------------
+      if prompts is None:
+        prompts = ["A dog running in a field with butterflies and tall grass"]
```
Comment (eltsai) [inline src/maxdiffusion/pipelines/flux/flux2klein_pipeline.py]: let's not hardcode the prompt here, we can read it from a yaml file (same to other prompts that are hard coded)

File: src/maxdiffusion/generate_flux2klein.py [hunk 8]
```diff
@@ -195,8 +221,13 @@ def main(argv):
   # 4. Load Qwen3 Config & Setup model layout
   from transformers import AutoConfig
 
-  max_logging.log(f"Loading Qwen3 config from text_encoder path: {text_encoder_path}...")
-  pt_config = AutoConfig.from_pretrained(text_encoder_path, local_files_only=True)
+  try:
+    pt_config = AutoConfig.from_pretrained(text_encoder_path, local_files_only=True)
+  except Exception:
+    depth_val = getattr(config, "depth", 24)
+    hf_repo = "black-forest-labs/FLUX.2-klein-9B" if depth_val in (24, -1) else "black-forest-labs/FLUX.2-klein-4B"
```
Comment (eltsai) [inline src/maxdiffusion/generate_flux2klein.py]: why are we doing this?

File: src/maxdiffusion/pipelines/flux/flux2klein_pipeline.py (hunk 6 above)
Comment (amepas) [reply src/maxdiffusion/pipelines/flux/flux2klein_pipeline.py]: Ah my mistake - was working on a separate PR which would keep the tokenizer here. I'll remove this change from this PR and add it to the next one

File: src/maxdiffusion/generate_flux2klein.py (hunk 5 above)
Comment (amepas) [reply src/maxdiffusion/generate_flux2klein.py:548]: Fixed!

File: src/maxdiffusion/generate_flux2klein.py [hunk 9]
```diff
@@ -134,15 +148,28 @@ def main(argv):
...
       load_and_convert_vae_weights,
-      cast_dict_to_bfloat16_inplace,
   )
   from maxdiffusion.pipelines.flux.flux2klein_pipeline import FlaxFlux2KleinPipeline
 
   config = pyconfig.config
   os.makedirs(config.output_dir, exist_ok=True)
 
+  if hasattr(config, "per_device_batch_size") and config.per_device_batch_size > 0:
+    calculated_batch_size = int(config.per_device_batch_size * jax.device_count())
```
Comment (eltsai) [inline src/maxdiffusion/generate_flux2klein.py:160]: maybe add a check/assertion here to avoid `calculated_batch_size=0` case, like if per_device_batch_size=0.25 and device count = 2, we will have int(0.25 * 2) == 0

File: src/maxdiffusion/tests/images/ref_flux2klein_9b.png
Comment (eltsai) [inline src/maxdiffusion/tests/images/ref_flux2klein_9b.png:1]: why did the ref image change? I did a run on the two commits and got `ssim(old_ref, new_ref, channel_axis=-1, data_range=255)  -> 0.7981`

Comment (amepas) [reply src/maxdiffusion/tests/images/ref_flux2klein_9b.png:1]: The ref image we had in the initial commit for the Flux 2 Klein model was stale (the implementation from the first commit would only give a 0.85 SSIM against it). The original reference image was generated using FP32 for the Qwen3 text encoder, whereas the committed code on maxdiffusion would only use bfloat16 (hence causing 1->0.85 SSIM drop). 

There was also an issue with how batch sizes were being handled in the initial commit, where the smoke test would always generate multiple images. This commit fixes that issue and generates only a single image. The SSIM dropped from (0.85->0.798) from this change. This drop was because XLA will fuse operations differently when we use a different batch size, and hence cause slight rounding errors in bfloat16 to accumulate and lead to a slightly lower SSIM.

File: src/maxdiffusion/generate_flux2klein.py (hunk 9 above)
Comment (amepas) [reply src/maxdiffusion/generate_flux2klein.py:160]: Resolved!

File: src/maxdiffusion/pipelines/flux/flux2klein_pipeline.py (hunk 7 above)
Comment (amepas) [reply src/maxdiffusion/pipelines/flux/flux2klein_pipeline.py]: Removed!

File: src/maxdiffusion/tests/images/ref_flux2klein_9b.png
Comment (amepas) [reply src/maxdiffusion/tests/images/ref_flux2klein_9b.png:1]: Marking this as resolved, please let me know if you have any other questions about this though

File: src/maxdiffusion/generate_flux2klein.py (hunk 8 above)
Comment (amepas) [reply src/maxdiffusion/generate_flux2klein.py]: This was an old check, no longer needed. Updated the code and simplified the other checks when loading the model configs from local or from HF.

**Review by eltsai — DISMISSED** [review]

File: src/maxdiffusion/configs/base_flux2klein.yml [hunk 10]
```diff
@@ -231,6 +231,7 @@ do_classifier_free_guidance: True
 guidance_scale: 4.0
 guidance_rescale: 0.0
 num_inference_steps: 4
+num_reps: 1
```
Comment (entrpn) [inline src/maxdiffusion/configs/base_flux2klein.yml:234]: why introduce a new variable instead of modifying num_inference_steps?

File: src/maxdiffusion/configs/base_flux2klein.yml [hunk 11]
```diff
@@ -203,7 +203,7 @@ num_train_epochs: 1
 seed: 0
 output_dir: 'output/'
 output_name: "flux2klein_generated_image.png"
-per_device_batch_size: 1
+per_device_batch_size: 1.0
```
Comment (entrpn) [inline src/maxdiffusion/configs/base_flux2klein.yml:206]: does this affect other configs that are using 1 instead of 1.0. In other words, does this affect how the config reads this number as an int or float?

**Review by entrpn — COMMENTED** [review]
left some comments.

File: src/maxdiffusion/configs/base_flux2klein.yml (hunk 10 above)
Comment (amepas) [reply src/maxdiffusion/configs/base_flux2klein.yml:234]: num_reps is the number of times we re-run inference after compilation - this was helpful in being able to report latency as an average over multiple runs instead of a single run. 

num_inference_steps is the number of denoising steps, which should be kept at 4 to maintain image quality.

File: src/maxdiffusion/configs/base_flux2klein.yml (hunk 11 above)
Comment (amepas) [reply src/maxdiffusion/configs/base_flux2klein.yml:206]: Yes, this should be fixed on other configs in this repo as well:

If we leave the definition in the .yml file as an integer, then the user cannot pass in floats as command line arguments since the config parses requires any command line arguments to have the same type as the variable defined in the .yml.

**Review by entrpn — APPROVED** [review]

**Review by eltsai — APPROVED** [review]

**Review by chandrasekhard2 — APPROVED** [review]

**Comment by amepas** [comment]
I am closing this Pull Request.  PR #458 included all changes in this branch and got merged first, so there is no need to merge this PR now!
